### PR TITLE
Add Emu workflow support part 2 

### PR DESCRIPTION
**Pull request by geishm-ansto**
**Description of work.**
Adds changes needed to support the EMU workflow locally at ANSTO. The original pull request request was closed to assess potential issues when some unrelated test modules failed the unit tests. 

**Report to:** Geish Miladinovic 

**To test:**

Existing unit test covers the loader.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by geishm-ansto** [comment]
Oops, I shouldn't have used the hash to reference the test file number as it is linked to a pull request.

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 1]
```diff
@@ -18,27 +18,31 @@
...
 #include <Poco/AutoPtr.h>
 #include <Poco/TemporaryFile.h>
 #include <Poco/Util/PropertyFileConfiguration.h>
 
-#include <math.h>
+#include <algorithm>
+#include <cmath>
+#include <cstdio>
+#include <fstream>
+#include <iostream>
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEMU.cpp]: We try to keep out `iostream` where possible. Is this required?

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 2]
```diff
@@ -231,8 +268,67 @@ class ConvertTOF {
     double t = invert(tsec, tn, t0);
     double tof = m_L0 / v1(t) + m_L2[detID] / m_v2;
 
-    return tof * 1.0e6;
+    return TofData(t * 1.0e6, tof * 1.0e6);
+  }
+};
+
+// calculate mean of a subset of the vector
+double maskedMean(std::vector<double> &vec, std::vector<bool> &mask) {
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEMU.cpp]: Can these both be `const std::vector<double> & `?

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 3]
```diff
@@ -231,8 +268,67 @@ class ConvertTOF {
...
+    sum += vec[i];
+    count++;
   }
+  if (count == 0)
+    throw std::runtime_error("mean of empty vector");
+  return sum / static_cast<double>(count);
+}
+
+// calculate stdev fro a subset of the vector
+double maskedStdev(std::vector<double> &vec, std::vector<bool> &mask) {
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEMU.cpp]: Can these both be `const std::vector<double> &`

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 4]
```diff
@@ -672,17 +811,20 @@ void LoadEMU::setupDetectorMasks(std::vector<bool> &roi) {
...
 
-// prepare the event storage
-void LoadEMU::prepareEventStorage(ANSTO::ProgressTracker &progTracker,
-                                  std::vector<size_t> &eventCounts,
-                                  std::vector<EventVector_pt> &eventVectors) {
+/// Allocate space for the event storage in \p eventVectors after the
+/// \p eventCounts have been determined.
+template <typename FD>
+void LoadEMU<FD>::prepareEventStorage(
+    ANSTO::ProgressTracker &progTracker, std::vector<size_t> &eventCounts,
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEMU.cpp]: `const std::vector &` here too?

File: Framework/DataHandling/src/LoadEMU.cpp [hunk 5]
```diff
@@ -717,11 +875,135 @@ void LoadEMU::loadDopplerParameters(API::LogManager &logm) {
...
+  // problem adding 'bool' to log
+  int32_t calPhase = (m_calibrateDoppler ? 1 : 0);
+  logm.addProperty("CalibratePhase", calPhase);
+}
+
+/// Calibrate the doppler phase based on the analysed events using
+/// the \p eventCounts and \p eventVectors.
+template <typename FD>
+void LoadEMU<FD>::calibrateDopplerPhase(
+    std::vector<size_t> &eventCounts,
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadEMU.cpp]: `const` vectors?

File: Framework/DataObjects/test/EventListTest.h [hunk 6]
```diff
@@ -1314,6 +1314,41 @@ class EventListTest : public CxxTest::TestSuite {
     }
   }
 
+  //-----------------------------------------------------------------------------------------------
+  void test_maskCondition_allTypes() {
```
Comment (martyngigg) [inline Framework/DataObjects/test/EventListTest.h]: :+1: for new tests

File: Framework/Geometry/src/Objects/MeshObject2D.cpp [hunk 7]
```diff
@@ -207,10 +206,11 @@ bool MeshObject2D::isValid(const Kernel::V3D &point) const {
 
   static const double tolerance = 1e-9;
   if (distanceToPlane(point) < tolerance) {
-    for (size_t i = 0; i < m_vertices.size(); i += 3) {
+    for (size_t i = 0; i < m_triangles.size(); i += 3) {
```
Comment (martyngigg) [inline Framework/Geometry/src/Objects/MeshObject2D.cpp]: Thanks for catching this!

File: Framework/PythonInterface/mantid/api/src/Exports/IEventList.cpp [hunk 8]
```diff
@@ -67,6 +73,8 @@ void export_IEventList() {
       .def("maskTof", &IEventList::maskTof, args("self", "tofMin", "tofMax"),
            "Mask out events that have a tof between tofMin and tofMax "
            "(inclusively)")
+      .def("maskCondition", &maskCondition, args("self", "mask"),
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/IEventList.cpp]: Can we add some tests for the Python export behaviour? Maybe in [`EventListTest`](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/test/python/mantid/dataobjects/EventListTest.py)?

**Review by martyngigg — CHANGES_REQUESTED** [review]
The changes look good. A few minor comments.

File: Framework/DataHandling/src/LoadEMU.cpp (hunk 1 above)
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: No it isn't necessary, I've removed it.

File: Framework/DataHandling/src/LoadEMU.cpp (hunk 2 above)
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Yes, changed.

File: Framework/DataHandling/src/LoadEMU.cpp (hunk 3 above)
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Yes, changed.

File: Framework/DataHandling/src/LoadEMU.cpp (hunk 4 above)
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Yes, changed.

File: Framework/DataHandling/src/LoadEMU.cpp (hunk 5 above)
Comment (geishm-ansto) [reply Framework/DataHandling/src/LoadEMU.cpp]: Yes, changed - apologies for all these missing consts.

File: Framework/PythonInterface/mantid/api/src/Exports/IEventList.cpp (hunk 8 above)
Comment (geishm-ansto) [reply Framework/PythonInterface/mantid/api/src/Exports/IEventList.cpp]: Added a test to the python event list test file.

**Comment by geishm-ansto** [comment]
@martyngigg Thanks for the feedback, I've completed the changes but the Ubuntu + Doc Tests failure is again not related to my work. I ran the KafkaHistoStreamDecoderTest that failed locally on Windows without any issues.

**Comment by martyngigg** [comment]
@geishm-ansto Thanks. The Kafka test seems to fail sporadically on the ubuntu build machines but we can't seem to reproduce it locally.

**Review by martyngigg — APPROVED** [review]
Code changes look good now :+1: 

The Kafka test failure is unrelated.

**Pull request merged**